### ISSUE TITLE
Small mistake in the combined queue variant of the swapchain acquire-present synchronization2 wiki example

### ISSUE THREAD

**Issue by miolad** [KhronosGroup/Vulkan-Docs#1919]
There is a small mistake is the `synchronization2` wiki for the combined queues variant of the `Swapchain Image Acquire and Present` example:

the rendering queue is submitted with a signal semaphore set to stage `VK_PIPELINE_STAGE_2_COLOR_ATTACHMENT_OUTPUT_BIT_KHR`, which should instead probably be `VK_PIPELINE_STAGE_2_ALL_COMMANDS_BIT_KHR` to chain with the second implicit external subpass dependency of the render pass and to mimic the behavior of a non-`synchronization2` implementation, which doesn't offer a `stageMask` parameter for the signal semaphore in a queue submit operation.

As of right now, the implicit external subpass dependency and the semaphore signal operation don't form a dependency chain, meaning that the automatic image layout transition into `finalLayout` could happen _after_ the `renderingCompleteSemaphore` is signaled.

**Comment by oddhack** [comment]
Could you link to the wiki page in question? This is the spec repo, and the spec isn't published in wiki form, so I'm not sure what you're talking about.

**Comment by miolad** [comment]
https://github.com/KhronosGroup/Vulkan-Docs/wiki/Synchronization-Examples#combined-graphicspresent-queue

It's this repo's wiki

**Comment by oddhack** [comment]
@tobski it looks like you were the last one to edit it - can you take this?

**Comment by Tobski** [comment]
This example is correct, it doesn't need to form a dependency chain with the subpass dependency - it just needs to happen before the load/store operations for the color attachments, which have to happen in VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT, so the condition is satisfied.

In fact actually looking at it, we don't need the subpass dependency for the first transition either, in either sync1 or sync2, and I'm now inclined to omit it. This is only really obvious since we added STAGE_NONE which is probably how it went unnoticed for so long.. I'm also now fairly sure these implicit dependencies should just be removed from the spec 🤔

**Comment by Tobski** [comment]
Oh the layout transitions mess things up don't they, so ok, that's why there's a first dependency here, so that can't be omitted, and also that's why it's in the spec. Ugh, ok so yes there does need to be a modification here, but I'll insert a second external dependency rather than over-synching the semaphore ops.

**Comment by Tobski** [comment]
Fixed in the latest update: https://github.com/KhronosGroup/Vulkan-Docs/wiki/Synchronization-Examples

@miolad Please close the issue if you are happy with the resolution.

**Comment by krOoze** [comment]
@Tobski I am not sure that is best. Oversyncing semaphores is how they generally work and how it workes in Vulkan 1.0. If the user wanted something not be covered by the semaphore, he could put it in another command buffer not covered by the semaphore.

But adding external dependency means there are now **two** synchronization primitives, where previously only one was needed. So it is always suboptimal.

**Comment by miolad** [comment]
@Tobski Yep, looks good, thanks. Especially because it's now exactly symmetrical w.r.t. to the acquire semaphore and first subpass dependency.

> Besides, the external dependency has almost the same problem as semaphore; it over-synchronizes everything after it.

@krOoze: isn't this the exact same kind of over-synchronization that's going on in reverse at the beginning of the render pass though? The subpass can't perform the layout transition (and draw operations enter their COLOR_ATTACHMENT_OUTPUT stage) until every prior operation on that queue has completed its COLOR ATTACHMENT_OUTPUT stage themselves, even if totally unrelated.

About this, from someone who is just getting started with Vulkan, it feels... weird not being able to restrict the synchronization scopes arbitrarily (apart from intra-render pass subpass dependencies, we can only express barriers in terms of "everything that comes before/after" depending on wether it's the first or second synchronization scope).
Is there a reason for this?

**Comment by Tobski** [comment]
Subpass dependencies skew this a lot. This would be a lot cleaner using dynamic rendering, then it'd be "barrier, render, barrier" with transitions done in the barriers. Really there's always some sort of sync going on, it's just subtle and allegedly implicit. There's no extra primitives here really, we're just asking the app to specify one to do a specific thing. We're just giving the implementation the chance to do something more optimal with non-fragment work if it can.

**Comment by krOoze** [comment]
@miolad Yes, but it is not supposed to be symmetric, because the situation is not symmetric; it is rolling-forward situation. You are usually synchronizing the outputs, which are (or should be) somewhat obviously at the end. In an incoming barrier you ask inputs to be ignored (geometry frontend), and postpone the semaphore until the output you are synchronizing. But outputs are at the end so it usually makes no sense to ask for the outgoing semaphore to be preponed, and is telling the driver something you potentially do not want. I usually view semaphore as a primitive that says "this batch is finished". It should be special circumstances when you instead want "this batch is finished except this and that".

Imagine you have properly batched work. You have one batch, with a semaphore saying that batch is finished. Then you have a second unrelated batch with a different semaphore saying that is finished. Now, if you add an unnecessery barrier, then you are saying that you want the second batch to actually depend on the first one. That feels like a parasitic sync. Blindly obeying driver should then prevent any overlap or synergy of that workload.

> we can only express barriers in terms of "everything that comes before/after" depending on wether it's the first or second synchronization scope).
Is there a reason for this?

Think of it as a regular processor with an operation stream, where you just put `MEM_BARR` somewhere in the operation stream depending on the stage parameters.

Now there are Vulkan Events for arbitrary point-to-point scopes. But that is not quite as trivial, isn't it? The processor now has an actual nontrivial object it needs to track the state of.

**Comment by Tobski** [comment]
@krOoze Yes but because there is a layout transition the "real" barrier is there anyway. I think you're overthinking this. I would concede that in this toy example it might be _slightly_ better to do it your way - but in all likelihood you're talking saving microseconds - just the cost of submitting an extra command. Sure, if it's not a toy example then yes, it wouldn't be worth putting it in there, unless you say, did some other work (e.g. compute) after the barrier (which admittedly means you'd need some extra barriers stuff in there).

More to the point though, the example as-is is correct, and I'm not sure there's much of note to be gained from doing it like sync1 forces you to instead; this way around makes it way more intuitive IMO.

**Comment by krOoze** [comment]
If there is layout transition then the "real" outgoing barrier has no reason to be there, if the semaphore does the job instead, otherwise there are two primitives doing the same. If the driver wants no layout transition there should be no barriers, and using the implicit dependency is exactly communicating that well.

It is not a toy example. It is an idiomatic example. Noobs will come look at this, and they will think it is the best possible option, and when doing their own thing, they will attempt to use identical logic to arrive at the best possible option for their situation.

It is a common question at community help sites why `pWaitDstStage` should be what it is. You need to provide equally compelling logic why `WaitSrcStage` should be what it is and "meh, it is just a toy example" is somewhat non-instructive.

>  I'm not sure there's much of note to be gained from doing it like sync1 forces you to instead

Consistency is king.

If sync2 can do it better, it is worth elaborating why and how it is so. Otherwise doing it the same way is the path of least surprise.

**Comment by miolad** [comment]
> Imagine you have properly batched work. You have one batch, with a semaphore saying that batch is finished. Then you have a second unrelated batch with a different semaphore saying that is finished. Now, if you add an unnecessery barrier, then you are saying that you want the second batch to actually depend on the first one. That feels like a parasitic sync. Blindly obeying driver should then prevent any overlap or synergy of that workload.

@krOoze if that second batch has an entry barrier to perform an image layout transition - like the first external subpass dependency in this case - then it is already needlessly dependent on the first batch, unless the `pWaitDstStage` of an hypothetical semaphore to the second batch is set to ALL_COMMANDS, in which case the barrier for the layout transition can use a `srcStageMask` of NONE.

I'm not sure why the second subpass dependency would be any worse than that.

**Comment by krOoze** [comment]
@miolad It is not worse, it is somewhat same bad, except twice now.

Still the difference is that it is somewhat output-oriented, and outputs are typically at the end. In `pWaitDstStage`+source dependency you order something to be finished and then something else can happily consume that output. In case of `WaitSrcStage`+destination dependency, you block something that would like to consume some previous output, but is prevented from doing so by the barrier. So, if your first batch is blocked by something, then transitionally the second batch is also blocked. I.e. it is possible the GPU does completely nothing, while it could, which is worse outcome than if GPU at least does something but it is demanded to finish it before starting something else.

**Comment by Tobski** [comment]
My expectation here is that between the two options, it's probably a moot point in terms of driver impact - I don't see any reason why a driver would go faster or slower by any significant amount with either method.

My justification for doing it the way I have is that explaining why the implicit dependency chain works has been a constant source of confusion for basically everyone I've ever discussed it with - putting in an explicit subpass dependency it makes it substantially clearer what is going on to a naive reader. That's my justification.

**Comment by miolad** [comment]
Since the original issue was fixed, and it seems like there is no further discussion here, I'm going to close this.

Thanks to everyone who participated!

**Comment by jbmcgee** [comment]
Came to back to this doc to verify my understanding of synchronization and saw this portion was changed. I then searched to understand why and found this thread. I think there is a loss in the changes and the previous version added more value. A lot of people come to the doc not for toy examples but to try to **understand** the synchronization aspects of the API with examples in their simplest form, which includes implicit transitions and `pWaitSemaphores`. Many people will think the second dependency is required and miss understanding some fundamentals of the API. If the previous version isn't preferred, I think at least mentioning the second pass isn't needed would go a long ways.

**Comment by Tobski** [comment]
> Came back to this doc to verify my understanding of synchronization and saw this portion was changed. I then searched to understand why and found this thread. I think there is a loss in the changes and the previous version added more value. A lot of people come to the doc not for toy examples but to try to **understand** the synchronization aspects of the API with examples in their simplest form, which includes implicit transitions and `pWaitSemaphores`. Many people will think the second dependency is required and miss understanding some fundamentals of the API. If the previous version isn't preferred, I think at least mentioning the second dependency isn't needed would go a long ways.

I don't think that implicit transitions are "the simplest form", they were a fudge we added to the API to make it work before we really understood the synchronization API we'd created, and generally my advice to everyone is to avoid relying on them, which is why I made the change I did. I'd rather people add in unnecessary explicit dependencies than flail around trying to understand the implicit ones - explicit dependencies add a lot of clarity in terms of readability that the implicit ones simply miss.  I've had countless conversations about how implicit dependencies work and _every time_ there's a fundamental misunderstanding of what they do.

> If the previous version isn't preferred, I think at least mentioning the second dependency isn't needed would go a long ways.

I'm not opposed to adding a note, so I can do that.

**Comment by Tobski** [comment]
> I'm not opposed to adding a note, so I can do that.

Done.
